modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java

```diff
@@ -990,6 +994,37 @@ public CompletableFuture<Ignite> start(Path configPath) {
 
                         return cmgMgr.onJoinReady();
                     }, startupExecutor)
+                    .thenComposeAsync(ignored -> {
+                        CompletableFuture<Void> awaitSelfInLogicalTopologyFuture = new CompletableFuture<>();
+
+                        LogicalTopologyEventListener awaitSelfListener = new LogicalTopologyEventListener() {
+                            @Override
+                            public void onNodeJoined(LogicalNode joinedNode, LogicalTopologySnapshot newTopology) {
+                                if (newTopology.nodes().stream().map(LogicalNode::id).collect(Collectors.toSet()).contains(id())) {
+                                    awaitSelfInLogicalTopologyFuture.complete(null);
+                                    logicalTopologyService.removeEventListener(this);
+                                }
+                            }
+
+                            @Override
+                            public void onTopologyLeap(LogicalTopologySnapshot newTopology) {
+                                if (newTopology.nodes().stream().map(LogicalNode::id).collect(Collectors.toSet()).contains(id())) {
```

Comment:
It seems like there are 3 duplicates of this code, it it possible to extract them to methods?

Reply:
Don't think that it's really needed, taking into considiration that everething inside thenCompose will be extracted I'd rather left inner duplication as is. However if you will intend I'll also add a private helper for this.

Reply:
Ok, will add inner method.